src/confirm.rs

```diff
@@ -1,48 +1,230 @@
-    .and_then(move |hash| {
-      wait_for_confirmations(transport.clone(), poll_interval, confirmations, move || {
-        eth.transaction_receipt(hash.clone()).map(|option| option.map(|receipt| receipt.block_number))
-      })
-      .and_then(move |_| {
-        let eth = Eth::new(transport.clone());
-        eth.transaction_receipt(hash).and_then(|option| option.ok_or(Error::Unreachable))
+pub trait ConfirmationCheck {
+  type Check: IntoFuture<Item = Option<U256>, Error = Error>;
+
+  fn check(&self) -> Self::Check;
+}
+
+enum WaitForConfirmationsState<F, O> {
+  WaitForNextBlock,
+  CheckConfirmation(F),
+  CompareConfirmations(u64, CallResult<U256, O>),
+}
+
+struct WaitForConfirmations<T, V, F> where T: Transport {
+  transport: T,
+  state: WaitForConfirmationsState<F, T::Out>,
+  filter_stream: Skip<FilterStream<T, H256>>,
+  confirmation_check: V,
+  confirmations: u64,
+}
+
+impl<T, V, F> Future for WaitForConfirmations<T, V, F::Future> where
+  T: Transport,
+  V: ConfirmationCheck<Check = F>,
+  F: IntoFuture<Item = Option<U256>, Error = Error> {
+
+  type Item = ();
+  type Error = Error;
+
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    loop {
+      let next_state = match self.state {
+        WaitForConfirmationsState::WaitForNextBlock => {
+          let _ = try_ready!(self.filter_stream.poll());
+          WaitForConfirmationsState::CheckConfirmation(self.confirmation_check.check().into_future())
+        },
+        WaitForConfirmationsState::CheckConfirmation(ref mut future) => match try_ready!(future.poll()) {
+          Some(confirmation_block_number) => {
+            let future = Eth::new(&self.transport).block_number();
+            WaitForConfirmationsState::CompareConfirmations(confirmation_block_number.low_u64(), future)
+          },
+          None => WaitForConfirmationsState::WaitForNextBlock,
+        },
+        WaitForConfirmationsState::CompareConfirmations(confirmation_block_number, ref mut block_number_future) => {
+          let block_number = try_ready!(block_number_future.poll()).low_u64();
+          if confirmation_block_number + self.confirmations >= block_number {
+            return Ok(().into())
+          } else {
+            WaitForConfirmationsState::WaitForNextBlock
+          }
+        },
+      };
+      self.state = next_state;
+    }
+  }
+}
+
+struct CreateWaitForConfirmations<T: Transport, V> {
+  create_filter: CreateFilter<T, H256>,
+  poll_interval: Duration,
+  transport: Option<T>,
+  confirmation_check: Option<V>,
+  confirmations: u64,
+}
+
+enum ConfirmationsState<T: Transport, V, F> {
+  Create(CreateWaitForConfirmations<T, V>),
+  Wait(WaitForConfirmations<T, V, F>),
+}
+
+pub struct Confirmations<T: Transport, V, F> {
+  state: ConfirmationsState<T, V, F>,
+}
+
+impl<T: Transport + Clone, V, F> Confirmations<T, V, F> {
+  fn new(transport: T, poll_interval: Duration, confirmations: u64, check: V) -> Self {
+    let eth = EthFilter::new(transport.clone());
```

Comment:
I think this could be just `&transport` here (and `+Clone` bound would be redundant)